(ch:strings-part3)=
# 문자열 3부

## 문자열과 for 반복문

준비중 ...

## 정규표현식

### Lesson 1: An Introduction, and the ABCs

Regular expressions are extremely useful in extracting information from text such as code, log files, spreadsheets, or even documents. And while there is a lot of theory behind formal languages, the following lessons and examples will explore the more practical uses of regular expressions so that you can use them as quickly as possible.

The first thing to recognize when using regular expressions is that everything is essentially a character, and we are writing patterns to match a specific sequence of characters (also known as a string). Most patterns use normal ASCII, which includes letters, digits, punctuation and other symbols on your keyboard like %#$@!, but unicode characters can also be used to match any type of international text.

Below are a couple lines of text, notice how the text changes to highlight the matching characters on each line as you type in the input field below. To continue to the next lesson, you will need to use the new syntax and concept introduced in each lesson to write a pattern that matches all the lines provided.

Go ahead and try writing a pattern that matches all three rows, it may be as simple as the common letters on each line.

In [1]:
import re

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `abcdefg` |
| 성사 | `abcde` |
| 성사 | `abc` |

답: `abc`

In [4]:
re.search('abc', 'abcdefg')

<re.Match object; span=(0, 3), match='abc'>

In [5]:
re.search('abc', 'abcde')

<re.Match object; span=(0, 3), match='abc'>

In [6]:
re.search('abc', 'abc')

<re.Match object; span=(0, 3), match='abc'>

### Lesson 1½: The 123s

Characters include normal letters, but digits as well. In fact, numbers 0-9 are also just characters and if you look at an ASCII table, they are listed sequentially.

Over the various lessons, you will be introduced to a number of special metacharacters used in regular expressions that can be used to match a specific type of character. In this case, the character \d can be used in place of any digit from 0 to 9. The preceding slash distinguishes it from the simple d character and indicates that it is a metacharacter.

Below are a few more lines of text containing digits. Try writing a pattern that matches all the digits in the strings below, and notice how your pattern matches anywhere within the string, not just starting at the first character. We will learn how to control this in a later lesson.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `abc123xyz` |
| 성사 | `deine "123"` |
| 성사 | `var g = 123;` |

- 숫자 활용

In [59]:
re.search('123', 'abc123xyz')

<re.Match object; span=(3, 6), match='123'>

In [61]:
re.search(r'\d', 'define "456789"')

<re.Match object; span=(8, 9), match='4'>

In [62]:
re.search(r'\d+', 'var g = 123;')

<re.Match object; span=(8, 11), match='123'>

- `\d` 활용

In [34]:
re.search(r'\d', 'abc123xyz')

<re.Match object; span=(3, 4), match='1'>

In [35]:
re.search(r'\d', 'define "123"')

<re.Match object; span=(8, 9), match='1'>

In [36]:
re.search(r'\d', 'var g = 123;')

<re.Match object; span=(8, 9), match='1'>

- `\D` 활용

In [63]:
re.search(r'\D*', 'abc123xyz')

<re.Match object; span=(0, 3), match='abc'>

In [38]:
re.search(r'\D', 'define "123"')

<re.Match object; span=(0, 1), match='d'>

In [39]:
re.search(r'\D', 'var g = 123;')

<re.Match object; span=(0, 1), match='v'>

### Lesson 2: The Dot

In some card games, the Joker is a wildcard and can represent any card in the deck. With regular expressions, you are often matching pieces of text that you don't know the exact contents of, other than the fact that they share a common pattern or structure (eg. phone numbers or zip codes).

Similarly, there is the concept of a wildcard, which is represented by the . (dot) metacharacter, and can match any single character (letter, digit, whitespace, everything). You may notice that this actually overrides the matching of the period character, so in order to specifically match a period, you need to escape the dot by using a slash \. accordingly.

Below are a couple strings with varying characters but the same length. Try to write a single pattern that can match the first three strings, but not the last (to be skipped). You may find that you will have to escape the dot metacharacter to match the period in some of the lines.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `cat.` |
| 성사 | `896.` |
| 성사 | `?=+.` |
| 미성사 | `abc1` |

In [40]:
re.search('.', 'cat.')

<re.Match object; span=(0, 1), match='c'>

In [41]:
re.search('.', '896.')

<re.Match object; span=(0, 1), match='8'>

In [44]:
re.search('.', '?=+.')

<re.Match object; span=(0, 1), match='?'>

In [45]:
re.search(r'\.', 'abc1')

In [20]:
re.search(r'.{3}\.', 'cat.')

<re.Match object; span=(0, 4), match='cat.'>

In [21]:
re.search(r'.{3}\.', '896.')

<re.Match object; span=(0, 4), match='896.'>

In [22]:
re.search(r'.{3}\.', '?=+.')

<re.Match object; span=(0, 4), match='?=+.'>

In [43]:
re.search(r'.{3}\.', 'abc1')

### Lesson 3: Matching specific characters

The dot metacharacter from the last lesson is pretty powerful, but sometimes too powerful. If we are matching phone numbers for example, we don't want to validate the letters "(abc) def-ghij" as being a valid number!

There is a method for matching specific characters using regular expressions, by defining them inside square brackets. For example, the pattern [abc] will only match a single a, b, or c letter and nothing else.

Below are a couple lines, where we only want to match the first three strings, but not the last three strings. Notice how we can't avoid matching the last three strings if we use the dot, but have to specifically define what letters to match using the notation above.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `can` |
| 성사 | `man` |
| 성사 | `fan` |
| 미성사 | `dan` |
| 미성사 | `ran` |
| 미성사 | `pan` |

In [46]:
re.search('[cmf]', 'can')

<re.Match object; span=(0, 1), match='c'>

In [47]:
re.search('[cmf]', 'man')

<re.Match object; span=(0, 1), match='m'>

In [48]:
re.search('[cmf]', 'fan')

<re.Match object; span=(0, 1), match='f'>

In [49]:
re.search('[cmf]an', 'dan')

In [50]:
re.search('[cmf]an', 'ran')

In [51]:
re.search('[cmf]an', 'pan')

### Lesson 4: Excluding specific characters

In some cases, we might know that there are specific characters that we don't want to match too, for example, we might only want to match phone numbers that are not from the area code 650.

To represent this, we use a similar expression that excludes specific characters using the square brackets and the ^ (hat). For example, the pattern [^abc] will match any single character except for the letters a, b, or c.

With the strings below, try writing a pattern that matches only the live animals (hog, dog, but not bog). Notice how most patterns of this type can also be written using the technique from the last lesson as they are really two sides of the same coin. By having both choices, you can decide which one is easier to write and understand when composing your own patterns.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `hog` |
| 성사 | `dog` |
| 미성사 | `bog` |

In [53]:
re.search('[^b]', 'hog')

<re.Match object; span=(0, 1), match='h'>

In [54]:
re.search('[^b]', 'dog')

<re.Match object; span=(0, 1), match='d'>

In [68]:
re.search('[^b]', 'bog')

<re.Match object; span=(1, 2), match='o'>

In [69]:
re.search('[^bo]', 'bog')

<re.Match object; span=(2, 3), match='g'>

In [70]:
re.search('[^bog]', 'bog')

### Lesson 5: Character ranges

We just learned how to create a pattern that matches or excludes specific characters -- but what if we want to match a character that can be in a sequential range characters? Do we have no choice but to list them all out?

Luckily, when using the square bracket notation, there is a shorthand for matching a character in list of sequential characters by using the dash to indicate a character range. For example, the pattern [0-6] will only match any single digit character from zero to six, and nothing else. And likewise, [^n-p] will only match any single character except for letters n to p.

Multiple character ranges can also be used in the same set of brackets, along with individual characters. An example of this is the alphanumeric \w metacharacter which is equivalent to the character range [A-Za-z0-9_] and often used to match characters in English text.

In the exercise below, notice how all the match and skip lines have a pattern, and use the bracket notation to match or skip each character from each line. Be aware that patterns are case sensitive and a-z differs from A-Z in terms of the characters it matches (lower vs upper case).

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `Ana` |
| 성사 | `boC` |
| 성사 | `Cpc` |
| 미성사 | `aax` |
| 미성사 | `bby` |
| 미성사 | `ccz` |

In [72]:
re.search('[A-C]', 'Ana')

<re.Match object; span=(0, 1), match='A'>

In [75]:
re.search('[A-C]', 'boC')

<re.Match object; span=(2, 3), match='C'>

In [76]:
re.search('[A-C]', 'Cpc')

<re.Match object; span=(0, 1), match='C'>

In [77]:
re.search('[A-C]', 'aax')

In [78]:
re.search('[A-C]', 'bby')

In [79]:
re.search('[A-C]', 'ccz')

In [81]:
re.search('[0-9]', '9ab')

<re.Match object; span=(0, 1), match='9'>

In [82]:
re.search('[0-9]', 'acb0')

<re.Match object; span=(3, 4), match='0'>

In [102]:
re.search(r'\w', 'acb0')

<re.Match object; span=(0, 1), match='a'>

In [103]:
re.search(r'\w', '\n')

In [105]:
re.search(r'\W', '\n')

<re.Match object; span=(0, 1), match='\n'>

In [106]:
re.search(r'\W', 'abc\n')

<re.Match object; span=(3, 4), match='\n'>

### Lesson 6: Catching some zzz's

Note: Some parts of the repetition syntax below isn't supported in all regular expression implementations.

We've so far learned how to specify the range of characters we want to match, but how about the number of repetitions of characters that we want to match? One way that we can do this is to explicitly spell out exactly how many characters we want, eg. \d\d\d which would match exactly three digits.

A more convenient way is to specify how many repetitions of each character we want using the curly braces notation. For example, a{3} will match the a character exactly three times. Certain regular expression engines will even allow you to specify a range for this repetition such that a{1,3} will match the a character no more than 3 times, but no less than once for example.

This quantifier can be used with any character, or special metacharacters, for example w{3} (three w's), [wxy]{5} (five characters, each of which can be a w, x, or y) and .{2,6} (between two and six of any character).

In the lines below, the last string with only one z isn't what we would consider a proper spelling of the slang "wazzup?". Try writing a pattern that matches only the first two spellings by using the curly brace notation above.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `wazzzzzup` |
| 성사 | `wazzzup` |
| 미성사 | `wazup` |

In [84]:
re.search('z{3,5}', 'wazzzzzup')

<re.Match object; span=(2, 7), match='zzzzz'>

In [88]:
re.search('z{3}', 'wazzzzzup')

<re.Match object; span=(2, 5), match='zzz'>

In [89]:
re.search('zzz', 'wazzzzzup')

<re.Match object; span=(2, 5), match='zzz'>

In [85]:
re.search('z{3,5}', 'wazzzup')

<re.Match object; span=(2, 5), match='zzz'>

In [86]:
re.search('z{3,5}', 'wazzup')

In [87]:
re.search('z{3,5}', 'wazup')

### Lesson 7: Mr. Kleene, Mr. Kleene

A powerful concept in regular expressions is the ability to match an arbitrary number of characters. For example, imagine that you wrote a form that has a donation field that takes a numerical value in dollars. A wealthy user may drop by and want to donate $25,000, while a normal user may want to donate $25.

One way to express such a pattern would be to use what is known as the Kleene Star and the Kleene Plus, which essentially represents either 0 or more or 1 or more of the character that it follows (it always follows a character or group). For example, to match the donations above, we can use the pattern \d* to match any number of digits, but a tighter regular expression would be \d+ which ensures that the input string has at least one digit.

These quantifiers can be used with any character or special metacharacters, for example a+ (one or more a's), [abc]+ (one or more of any a, b, or c character) and .* (zero or more of any character).

Below are a few simple strings that you can match using both the star and plus metacharacters.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `aaaabcc` |
| 성사 | `aabbbbc` |
| 성사 | `aacc` |
| 미성사 | `a` |

In [90]:
re.search('aa+', 'aaaabcc')

<re.Match object; span=(0, 4), match='aaaa'>

In [91]:
re.search('aa+', 'aabbbbc')

<re.Match object; span=(0, 2), match='aa'>

In [92]:
re.search('aa+', 'aacc')

<re.Match object; span=(0, 2), match='aa'>

In [93]:
re.search('aa+', 'a')

In [94]:
re.search('aa+', 'ab')

In [107]:
re.search('aa*', 'a')

<re.Match object; span=(0, 1), match='a'>

### Lesson 8: Characters optional

As you saw in the previous lesson, the Kleene star and plus allow us to match repeated characters in a line.

Another quantifier that is really common when matching and extracting text is the ? (question mark) metacharacter which denotes optionality. This metacharacter allows you to match either zero or one of the preceding character or group. For example, the pattern ab?c will match either the strings "abc" or "ac" because the b is considered optional.

Similar to the dot metacharacter, the question mark is a special character and you will have to escape it using a slash \? to match a plain question mark character in a string.

In the strings below, notice how the the plurality of the word "file" depends on the number of files found. Try writing a pattern that uses the optionality metacharacter to match only the lines where one or more files were found.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `1 file found?` |
| 성사 | `2 files found?` |
| 성사 | `24 files found?` |
| 미성사 | `No files found.` |

In [96]:
re.search(r'\d+.*\?', '1 file found?')

<re.Match object; span=(0, 13), match='1 file found?'>

In [97]:
re.search(r'\d+.*\?', '2 files found?')

<re.Match object; span=(0, 14), match='2 files found?'>

In [98]:
re.search(r'\d+.*\?', '24 files found?')

<re.Match object; span=(0, 15), match='24 files found?'>

In [99]:
re.search(r'\d+.*\?', 'No files found.')

In [110]:
re.search(r'\d\d?', '24 files found?')

<re.Match object; span=(0, 2), match='24'>

In [111]:
re.search(r'\d\d\d?', '24 files found?')

<re.Match object; span=(0, 2), match='24'>

### Lesson 9: All this whitespace

When dealing with real-world input, such as log files and even user input, it's difficult not to encounter whitespace. We use it to format pieces of information to make it easier to read and scan visually, and a single space can put a wrench into the simplest regular expression.

The most common forms of whitespace you will use with regular expressions are the space (␣), the tab (\t), the new line (\n) and the carriage return (\r) (useful in Windows environments), and these special characters match each of their respective whitespaces. In addition, a whitespace special character \s will match any of the specific whitespaces above and is extremely useful when dealing with raw input text.

In the strings below, you'll find that the content of each line is indented by some whitespace from the index of the line (the number is a part of the text to match). Try writing a pattern that can match each line containing whitespace characters between the number and the content. Notice that the whitespace characters are just like any other character and the special metacharacters like the star and the plus can be used as well.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `1. abc` |
| 성사 | `2.  abc` |
| 성사 | `3.   abc` |
| 미성사 | `4.abc` |

In [113]:
re.search(r'\d\.\s+abc', '1. abc')

<re.Match object; span=(0, 6), match='1. abc'>

In [114]:
re.search(r'\d\.\s+abc', '2.  abc')

<re.Match object; span=(0, 7), match='2.  abc'>

In [115]:
re.search(r'\d\.\s+abc', '3.   abc')

<re.Match object; span=(0, 8), match='3.   abc'>

In [116]:
re.search(r'\d\.\s+abc', '4.abc')

In [117]:
re.search(r'\d\.\S', '4.abc')

<re.Match object; span=(0, 3), match='4.a'>

### Lesson 10: Starting and ending

So far, we've been writing regular expressions that partially match pieces across all the text. Sometimes this isn't desirable, imagine for example we wanted to match the word "success" in a log file. We certainly don't want that pattern to match a line that says "Error: unsuccessful operation"! That is why it is often best practice to write as specific regular expressions as possible to ensure that we don't get false positives when matching against real world text.

One way to tighten our patterns is to define a pattern that describes both the start and the end of the line using the special ^ (hat) and $ (dollar sign) metacharacters. In the example above, we can use the pattern ^success to match only a line that begins with the word "success", but not the line "Error: unsuccessful operation". And if you combine both the hat and the dollar sign, you create a pattern that matches the whole line completely at the beginning and end.

Note that this is different than the hat used inside a set of bracket [^...] for excluding characters, which can be confusing when reading regular expressions.

Try to match each of the strings below using these new special characters.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `Mission: successful` |
| 미성사 | `Last Mission: unsuccessful` |
| 미성사 | `Next Mission: successful upon capture of target` |

In [118]:
re.search('^Mission.*ful$', 'Mission: successful')

<re.Match object; span=(0, 19), match='Mission: successful'>

In [119]:
re.search('^Mission.*ful$', 'Last Mission: unsuccessful')

In [120]:
re.search('^Mission.*ful$', 'Next Mission: successful upon capture of target')

### Lesson 11: Match groups

Regular expressions allow us to not just match text but also to extract information for further processing. This is done by defining groups of characters and capturing them using the special parentheses ( and ) metacharacters. Any subpattern inside a pair of parentheses will be captured as a group. In practice, this can be used to extract information like phone numbers or emails from all sorts of data.

Imagine for example that you had a command line tool to list all the image files you have in the cloud. You could then use a pattern such as ^(IMG\d+\.png)$ to capture and extract the full filename, but if you only wanted to capture the filename without the extension, you could use the pattern ^(IMG\d+)\.png$ which only captures the part before the period.

Go ahead and try to use this to write a regular expression that matches only the filenames (not including extension) of the PDF files below.

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `file_record_transcript.pdf` | `file_record_transcript` |
| 성사 | `file_07241999.pdf` | `file_07241999` |
| 미성사 | `testfile_fake.pdf.tmp` | |

In [146]:
m = re.search(r'^file.+\.pdf$', 'file_record_trasncript.pdf')
m

<re.Match object; span=(0, 26), match='file_record_trasncript.pdf'>

In [147]:
m.group(0)

'file_record_trasncript.pdf'

In [148]:
m.group(1)

IndexError: no such group

In [149]:
mg = re.search(r'^(file.+)\.pdf$', 'file_record_trasncript.pdf')
mg

<re.Match object; span=(0, 26), match='file_record_trasncript.pdf'>

In [150]:
mg.group(0)

'file_record_trasncript.pdf'

In [151]:
mg.group(1)

'file_record_trasncript'

In [152]:
mg = re.search(r'^(file.+)\.(pdf)$', 'file_record_trasncript.pdf')
mg

<re.Match object; span=(0, 26), match='file_record_trasncript.pdf'>

In [153]:
mg.group(0)

'file_record_trasncript.pdf'

In [154]:
mg.group(1)

'file_record_trasncript'

In [155]:
mg.group(2)

'pdf'

In [156]:
mg.group(3)

IndexError: no such group

### Lesson 12: Nested groups

When you are working with complex data, you can easily find yourself having to extract multiple layers of information, which can result in nested groups. Generally, the results of the captured groups are in the order in which they are defined (in order by open parenthesis).

Take the example from the previous lesson, of capturing the filenames of all the image files you have in a list. If each of these image files had a sequential picture number in the filename, you could extract both the filename and the picture number using the same pattern by writing an expression like ^(IMG(\d+))\.png$ (using a nested parenthesis to capture the digits).

The nested groups are read from left to right in the pattern, with the first capture group being the contents of the first parentheses group, etc.

For the following strings, write an expression that matches and captures both the full date, as well as the year of the date.

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `Jan 1987` | `Jan 1987` `1987` |
| 성사 | `May 1969` | `May 1969` `1969` |
| 성사 | `Aug 2011` | `Aug 2011` `2011` |

In [158]:
mg = re.search(r'(\w+ (\d+))', 'Jan 1987')

In [160]:
mg.group(0)

'Jan 1987'

In [161]:
mg.group(1)

'Jan 1987'

In [162]:
mg.group(2)

'1987'

### Lesson 13: More group work

As you saw in the previous lessons, all the quantifiers including the star *, plus +, repetition {m,n} and the question mark ? can all be used within the capture group patterns. This is the only way to apply quantifiers on sequences of characters instead of the individual characters themselves.

For example, if I knew that a phone number may or may not contain an area code, the right pattern would test for the existence of the whole group of digits (\d{3})? and not the individual characters themselves (which would be wrong).

Depending on the regular expression engine you are using, you can also use non-capturing groups which will allow you to match the group but not have it show up in the results.

Below are a couple different common display resolutions, try to capture the width and height of each display.

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `1280x720` | `1280` `720` |
| 성사 | `1920x1600` | `1920` `1600` |
| 성사 | `024x768` | `1024` `768` |

In [164]:
mg = re.search(r'(\d+)x(\d+)', '1280x720')

In [166]:
mg.group(0)

'1280x720'

In [167]:
mg.group(1)

'1280'

In [168]:
mg.group(2)

'720'

In [171]:
mg = re.search(r'(\d+)x(\d+)', '1920x1600')

In [172]:
mg.group(0)

'1920x1600'

In [173]:
mg.group(1)

'1920'

In [174]:
mg.group(2)

'1600'

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `010-1234-5742` | `1234` `5742` |
| 성사 | `1234-5742` | `1234` `5742` |

In [186]:
mg = re.search(r'\d+-(\d+)-(\d+)', '010-1234-5678')
mg

<re.Match object; span=(0, 13), match='010-1234-5678'>

In [187]:
mg.group(0)

'010-1234-5678'

In [188]:
mg.group(1)

'1234'

In [189]:
mg.group(2)

'5678'

In [271]:
mg = re.search(r'((\d+)-)?(\d+)-(\d+)', '1234-5678')
mg

<re.Match object; span=(0, 9), match='1234-5678'>

In [272]:
mg.group(0)

'1234-5678'

In [273]:
mg.group(1)

In [274]:
mg.group(2)

In [275]:
mg.group(3)

'1234'

In [276]:
mg.group(4)

'5678'

In [264]:
mg = re.search(r'((\d+)-)?(\d+)-(\d+)', '010-1234-5678')
mg

<re.Match object; span=(0, 13), match='010-1234-5678'>

In [265]:
mg.group(0)

'010-1234-5678'

In [266]:
mg.group(1)

'010-'

In [267]:
mg.group(2)

'010'

In [268]:
mg.group(3)

'1234'

In [269]:
mg.group(4)

'5678'

### Lesson 14: It's all conditional

As we mentioned before, it's always good to be precise, and that applies to coding, talking, and even regular expressions. For example, you wouldn't write a grocery list for someone to Buy more .* because you would have no idea what you could get back. Instead you would write Buy more milk or Buy more bread, and in regular expressions, we can actually define these conditionals explicitly.

Specifically when using groups, you can use the | (logical OR, aka. the pipe) to denote different possible sets of characters. In the above example, I can write the pattern "Buy more (milk|bread|juice)" to match only the strings Buy more milk, Buy more bread, or Buy more juice.

Like normal groups, you can use any sequence of characters or metacharacters in a condition, for example, ([cb]ats*|[dh]ogs?) would match either cats or bats, or, dogs or hogs. Writing patterns with many conditions can be hard to read, so you should consider making them separate patterns if they get too complex.

Go ahead and try writing a conditional pattern that matches only the lines with small fuzzy creatures below.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `I love cats` |
| 성사 | `I love dogs` |
| 미성사 | `I love logs` |
| 미성사 | `I love cogs` |

In [278]:
mg = re.search('I love (cats|dogs)', 'I love cats')
mg

<re.Match object; span=(0, 11), match='I love cats'>

In [279]:
mg.group(0)

'I love cats'

In [280]:
mg.group(1)

'cats'

In [281]:
mg.group(2)

IndexError: no such group

In [283]:
mg = re.search('I love (cats|dogs)', 'I love dogs')
mg

<re.Match object; span=(0, 11), match='I love dogs'>

In [284]:
mg.group(0)

'I love dogs'

In [285]:
mg.group(1)

'dogs'

In [286]:
mg.group(2)

IndexError: no such group

In [287]:
mg = re.search('I love (cats|dogs)', 'I love logs')
mg

In [294]:
mg = re.search('I love ([cb]ats*|[dh]ogs?)', 'I love hogs')
mg

<re.Match object; span=(0, 11), match='I love hogs'>

In [295]:
mg.group(0)

'I love hogs'

In [296]:
mg.group(1)

'hogs'

### Problem 1: Matching a decimal numbers

At first glance, writing a regular expression to match a number should be easy right?

We have the \d special character to match any digit, and all we need to do is match the decimal point right? For simple numbers, that may be right, but when working with scientific or financial numbers, you often have to deal with positive and negative numbers, significant digits, exponents, and even different representations (like the comma used to separate thousands and millions).

Below are a few different formats of numbers that you might encounter. Notice how you will have to match the decimal point itself and not an arbitrary character using the dot metacharacter. If you are having trouble skipping the last number, notice how that number ends the line compared to the rest.

| 매칭 | 문자열 |
| :---: | :--- |
| 성사 | `3.14529` |
| 성사 | `-255.34` |
| 성사 | `128` |
| 성사 | `1.9e10` |
| 성사 | `123,340.00` |
| 미성사 | `720p` |

답:

In [298]:
re.search(r'^-?\d+(,\d+)*(\.\d+(e\d+)?)?$', '3.14529')

<re.Match object; span=(0, 7), match='2.14529'>

In [299]:
re.search(r'^-?\d+(,\d+)*(\.\d+(e\d+)?)?$', '-255.34')

<re.Match object; span=(0, 7), match='-255.34'>

In [300]:
re.search(r'^-?\d+(,\d+)*(\.\d+(e\d+)?)?$', '128')

<re.Match object; span=(0, 3), match='128'>

In [301]:
re.search(r'^-?\d+(,\d+)*(\.\d+(e\d+)?)?$', '1.9e10')

<re.Match object; span=(0, 6), match='1.9e10'>

In [302]:
re.search(r'^-?\d+(,\d+)*(\.\d+(e\d+)?)?$', '123,340.00')

<re.Match object; span=(0, 10), match='123,340.00'>

In [303]:
re.search(r'^-?\d+(,\d+)*(\.\d+(e\d+)?)?$', '720p')

### Problem 2: Matching phone numbers

Validating phone numbers is another tricky task depending on the type of input that you get. Having phone numbers from out of the state which require an area code, or international numbers which require a prefix will add complexity to the regular expression, as does the individual preferences that people have for entering phone numbers (some put dashes or whitespace while others do not for example).

Below are a few phone numbers that you might encounter when using real data, write a single regular expressions that matches the number and captures the proper area code.

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `415-555-1234` | `415` |
| 성사 | `650-555-2345` | `650` |
| 성사 | `(416)555-3456` | `416` |
| 성사 | `202 555 4567` | `202` |
| 성사 | `4035555678` | `403` |
| 성사 | `+82 416 555 9292` | `416` |

In [327]:
mg = re.search(r'(\+82)?[\s-]?\(?(\d{3})\)?[\s-]?\d{3}[\s-]?\d{4}', '415-555-1234')
mg

<re.Match object; span=(0, 12), match='415-555-1234'>

In [328]:
mg.group(2)

'415'

In [329]:
mg = re.search(r'(\+82)?[\s-]?\(?(\d{3})\)?[\s-]?\d{3}[\s-]?\d{4}', '(416)555-3456')
mg

<re.Match object; span=(0, 13), match='(416)555-3456'>

In [330]:
mg.group(2)

'416'

In [331]:
mg = re.search(r'(\+82)?[\s-]?\(?(\d{3})\)?[\s-]?\d{3}[\s-]?\d{4}', '202 555 4567')
mg

<re.Match object; span=(0, 12), match='202 555 4567'>

In [332]:
mg.group(2)

'202'

In [333]:
mg = re.search(r'(\+82)?[\s-]?\(?(\d{3})\)?[\s-]?\d{3}[\s-]?\d{4}', '4035555678')
mg

<re.Match object; span=(0, 10), match='4035555678'>

In [334]:
mg.group(2)

'403'

In [338]:
mg = re.search(r'(\+82)?[\s-]?\(?(\d{3})\)?[\s-]?\d{3}[\s-]?\d{4}', '+82 416 555 9292')
mg

<re.Match object; span=(0, 16), match='+82 416 555 9292'>

In [339]:
mg.group(2)

'416'

### Problem 3: Matching emails

When you are dealing with HTML forms, it's often useful to validate the form input against regular expressions. In particular, emails are difficult to match correctly due to the complexity of the specification and I would recommend using a built-in language or framework function instead of rolling your own. However, you can build a pretty robust regular expression that matches a great deal of common emails pretty easily using what we've learned so far.

One thing to watch out for is that many people use plus addressing for one time use, such as "name+filter@gmail.com", which gets directly to "name@gmail.com" but can be filtered with the extra information. In addition, some domains have more than one component, for example, you can register a domain at "hellokitty.hk.com" and have an email with the form "ilove@hellokitty.hk.com", so you will have to be careful when matching the domain portion of the email.

Below are a few common emails, in this example, try to capture the name of the email, excluding the filter (+ character and afterwards) and domain (@ character and afterwards).

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `tom@hogwarts.com` | `tom` |
| 성사 | `tom.riddle@hogwarts.com` | `tom.riddle` |
| 성사 | `tom.riddle+regexone@hogwarts.com` | `tom.riddle` |
| 성사 | `tom@hogwarts.eu.com` | `tom` |
| 성사 | `potter@hogwarts.com` | `potter` |
| 성사 | `harry@hogwarts.com` | `harry` |
| 성사 | `hermione+regexone@hogwarts.com` | `hermione` |

In [340]:
pattern = re.compile(r'^([\w\.]+)')

In [343]:
mg = pattern.search('tom@hogwarts.com')
mg

<re.Match object; span=(0, 3), match='tom'>

In [353]:
mg.group(0)

'tom'

In [356]:
mg.group(1)

'tom'

In [357]:
mg = pattern.search('tom.riddle@hogwarts.com')
mg

<re.Match object; span=(0, 10), match='tom.riddle'>

In [358]:
mg.group(0)

'tom.riddle'

In [359]:
mg.group(1)

'tom.riddle'

In [360]:
mg = pattern.search('tom.riddle+regexone@hogwarts.com')
mg

<re.Match object; span=(0, 10), match='tom.riddle'>

In [361]:
mg.group(0)

'tom.riddle'

In [362]:
mg.group(1)

'tom.riddle'

In [364]:
mg = pattern.search('tom@hogwarts.eu.com')
mg

<re.Match object; span=(0, 3), match='tom'>

In [365]:
mg.group(0)

'tom'

In [366]:
mg.group(1)

'tom'

### Problem 4: Matching HTML

If you are looking for a robust way to parse HTML, regular expressions are usually not the answer due to the fragility of html pages on the internet today -- common mistakes like missing end tags, mismatched tags, forgetting to close an attribute quote, would all derail a perfectly good regular expression. Instead, you can use libraries like Beautiful Soup or html5lib (both Python) or phpQuery (PHP) which not only parse the HTML but allow you to walk to DOM quickly and easily.

That said, there are often times when you want to quickly match tags and tag content in an editor, and if you can vouch for the input, regular expressions are a good tool to do this. As you can see in the examples below, some things that you might want to be careful about odd attributes that have extra escaped quotes and nested tags.

Go ahead and write regular expressions for the following examples.

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `<a>This is a link</a>` | `a` |
| 성사 | `<a href='https://regexone.com'>Link</a>` | `a` |
| 성사 | `<div class='test_style'>Test</div>` | `div` |
| 성사 | `<div>Hello <span>world</span></div>` | `div` |

In [5]:
mg = re.search(r'<(\w+)', '<a>This is a link</a>')
mg

<re.Match object; span=(0, 2), match='<a'>

In [6]:
mg.group(0)

'<a'

In [7]:
mg.group(1)

'a'

In [12]:
mg = re.search(r'<(\w+)', "<a href='https://regexone.com'>Link</a>")
mg

<re.Match object; span=(0, 2), match='<a'>

In [13]:
mg.group(0)

'<a'

In [14]:
mg.group(1)

'a'

In [15]:
mg = re.search(r'<(\w+)', "<div class='test_style'>Test</div>")
mg

<re.Match object; span=(0, 4), match='<div'>

In [16]:
mg.group(0)

'<div'

In [17]:
mg.group(1)

'div'

In [18]:
mg = re.search(r'<(\w+)', "<div>Hello <span>world</span></div>")
mg

<re.Match object; span=(0, 4), match='<div'>

In [19]:
mg.group(0)

'<div'

In [20]:
mg.group(1)

'div'

In [21]:
mg = re.search(r'>([\w\s]*)<', '<a>This is a link</a>')
mg

<re.Match object; span=(2, 18), match='>This is a link<'>

In [22]:
mg.group(0)

'>This is a link<'

In [23]:
mg.group(1)

'This is a link'

In [28]:
mg = re.search(r'>([\w\s]*)<', "<a href='https://regexone.com'>Link</a>")
mg

<re.Match object; span=(30, 36), match='>Link<'>

In [29]:
mg.group(0)

'>Link<'

In [30]:
mg.group(1)

'Link'

In [31]:
mg = re.search(r'>([\w\s]*)<', "<div class='test_style'>Test</div>")
mg

<re.Match object; span=(23, 29), match='>Test<'>

In [32]:
mg.group(0)

'>Test<'

In [33]:
mg.group(1)

'Test'

In [37]:
mg = re.search(r'>([\w\s]*)<', "<div>Hello <span>world</span></div>")
mg

<re.Match object; span=(4, 12), match='>Hello <'>

In [38]:
mg.group(0)

'>Hello <'

In [39]:
mg.group(1)

'Hello '

In [58]:
mg = re.search(r"='([\w:/\.]*)'", "<a href='https://regexone.com'>Link</a>")
mg

<re.Match object; span=(7, 30), match="='https://regexone.com'">

In [59]:
mg.group(0)

"='https://regexone.com'"

In [60]:
mg.group(1)

'https://regexone.com'

In [61]:
mg = re.search(r"='([\w:/\.]*)'", "	<div class='test_style'>Test</div>")
mg

<re.Match object; span=(11, 24), match="='test_style'">

In [62]:
mg.group(0)

"='test_style'"

In [63]:
mg.group(1)

'test_style'

### Problem 5: Matching specific filenames

If you use Linux or the command line frequently, are often dealing with lists of files. Most files have a filename component as well as an extension, but in Linux, it is also common to have hidden files that have no filename.

In this simple example, extract the filenames and extension types of only image files (not including temporary files for images currently being edited). Image files are defined as .jpg,.png, and .gif.

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 미성사 | `.bash_profile` | |
| 미성사 | `workspace.doc` | |
| 성사 | `img0912.jpg` | `img0912` `jpg` |
| 성사 | `updated_img0912.png` | `updated_img0912` `png` |
| 미성사 | `documentation.html` | |
| 성사 | `favicon.gig` | `favicon` `gif` |
| 미성사 | `img0912.jpg.tmp` | |
| 미성사 | `access.lock` | |

In [3]:
mg = re.search(r'(\w+)\.(jpg|png|gif)$', '.bash_profile')
mg

In [4]:
mg = re.search(r'(\w+)\.(jpg|png|gif)$', 'workspace.doc')
mg

In [5]:
mg = re.search(r'(\w+)\.(jpg|png|gif)$', 'img0912.jpg')
mg

<re.Match object; span=(0, 11), match='img0912.jpg'>

In [8]:
mg.group(1)

'img0912'

In [9]:
mg.group(2)

'jpg'

In [10]:
mg = re.search(r'(\w+)\.(jpg|png|gif)$', 'updated_img0912.png')
mg

<re.Match object; span=(0, 19), match='updated_img0912.png'>

In [11]:
mg.group(1)

'updated_img0912'

In [12]:
mg.group(2)

'png'

### Problem 6: Trimming whitespace from start and end of line

Occasionally, you'll find yourself with a log file that has ill-formatted whitespace where lines are indented too much or not enough. One way to fix this is to use an editor's search a replace and a regular expression to extract the content of the lines without the extra whitespace.

We have previously seen how to match a full line of text using the hat ^ and the dollar sign $ respectively. When used in conjunction with the whitespace \s, you can easily skip all preceding and trailing spaces.

Write a simple regular expression to capture the content of each line, without the extra whitespace.

| 그룹화 | 문자열 | 그룹 |
| :---: | :--- | :--- |
| 성사 | `    The quick brown fox...` | `The quick brown fox...` |
| 성사 | `jumps over the lazy dog.    ` | `jumps over the lazy dog.` |

In [36]:
mg = re.search(r'^\s*(.*)\s*$', '    The quick brown fox...')
mg

In [25]:
mg.group(1)

'The quick brown fox...'

In [31]:
mg = re.search(r'^\s*(\w+\s?)*\s*$', '    jumps over the lazy dog.    ')
mg

In [27]:
mg.group(1)

'jumps over the lazy dog.    '

In [48]:
mg = re.search(r'^\s*((?:[\w\.]+\s)*[\w\.]+)\s*$', '    jumps over the lazy dog.    ')
mg

<re.Match object; span=(0, 32), match='    jumps over the lazy dog.    '>

In [49]:
mg.group(1)

'jumps over the lazy dog.'

정규표현식<font size='2'>regular expression</font>은
**패턴 탐색 문자열**이며 정규표현식(레겍스)라고 줄여서 불리기도 한다.
즉, 정규표현식은 주어진 문자열이 특정 패턴의 부분 문자열을 포함하는가를
판단할 때 주로 활용된다.

### 정규표현식 모듈

파이썬에 정규표현식을 활용하려면 `re` 모듈이 유용하다.

예를 들어 아래 코드는 `txt` 변수가 가리키는 문장이
`"The"`로 시자하면서 동시에 `"Korea"`으로 끝나는지 여부를 판단한다.

In [34]:
txt = "The rain in Korea"
m = re.search("^The.*Korea$", txt)

In [35]:
m.group(0)

'The rain in Korea'

In [41]:
txt = "The rain in Summer, and the rain in Autumn in Korea"
x = re.search("ai", txt)

In [45]:
x.start()

5

In [53]:
x.end()

7

In [54]:
x.span()

(5, 7)

In [55]:
x

<re.Match object; span=(5, 7), match='ai'>

In [42]:
x.group(0)

'ai'

In [52]:
m.groupdict()

{}

In [37]:
print(x) #this will print an object

<re.Match object; span=(5, 7), match='ai'>


### 정규표현식 Functions

The re module offers a set of functions that allows us to search a string for a match:

:::{list-table} 정규식 함수
:widths: 10 30
:header-rows: 1
:name: 정규표현식-functions

*   - 기능
    - 설명
*   - findall
    - Returns a list containing all matches
*   - search
    - Returns a Match object if there is a match anywhere in the string
*   - split
    - Returns a list where the string has been split at each match
*   - sub
    - Replaces one or many matches with a string
:::

### Metacharacters

Metacharacters are characters with a special meaning:

:::{list-table} 메타 기호
:widths: 5 25 10
:header-rows: 1
:name: metacharacters

*   - 기호
    - 설명
    - 예제
*   - `[]`
    - A set of characters
    - `"[a-m]"`
*   - `\`
    - Signals a special sequence (can also be used to escape special characters)
    - `"\d"`
*   - `.`
    - Any character (except newline character)
    - `"he..o"`    
*   - `^`
    - Starts with
    - `"^hello"`
*   - `$`
    - Ends with
    - `"planet$"`
*   - `*`
    - Zero or more occurrences
    - `"he.*o"`
*   - `+`
    - One or more occurrences
    - `"he.+o"`
*   - `?`
    - Zero or one occurrences
    - `"he.?o"`
*   - `{}`
    - Exactly the specified number of occurrences
    - `"he.{2}o"`
*   - `|`
    - Either or
    - `"falls|stays"`
*   - `()`
    - Capture and group
    - 
:::

### Special Sequences

A special sequence is a \ followed by one of the characters in the list below, and has a special meaning:

:::{list-table} 특별 기호
:widths: 5 25 10
:header-rows: 1
:name: special-seq

*   - 기호
    - 설명
    - 예제
*   - `\A`
    - Returns a match if the specified characters are at the beginning of the string
    - `"\AThe"`
*   - `\b`
    - Returns a match where the specified characters are at the beginning or at the end of a word
    - `r"\bain"`
*   - 
    - (the "r" in the beginning is making sure that the string is being treated as a "raw string")
    - `r"ain\b"`
*   - `\B`
    - Returns a match where the specified characters are present, but NOT at the beginning (or at the end) of a 
    - `r"\Bain"`
*   - 
    - word (the "r" in the beginning is making sure that the string is being treated as a "raw string")	
    - `r"ain\B"`
*   - `\d`
    - Returns a match where the string contains digits (numbers from 0-9)
    - `"\d"`
*   - `\D`
    - Returns a match where the string DOES NOT contain digits
    - `"\D"`
*   - `\s`
    - Returns a match where the string contains a white space character
    - `"\s"`
*   - `\S`
    - Returns a match where the string DOES NOT contain a white space character
    - `"\S"`
*   - `\w`
    - Returns a match where the string contains any word characters (characters from a to Z, digits from 0-9, and the underscore _ character)	
    - `"\w"`
*   - `\W`
    - Returns a match where the string DOES NOT contain any word characters
    - `"\W"`
*   - `\Z`
    - Returns a match if the specified characters are at the end of the string
    - `"Korea\Z"`
:::

### Sets

A set is a set of characters inside a pair of square brackets [] with a special meaning:

:::{list-table} 집합
:widths: 10 30
:header-rows: 1
:name: sets

*   - 집합
    - 설명
*   - `[arn]`
    - Returns a match where one of the specified characters (a, r, or n) is present
*   - `[a-n]`
    - Returns a match for any lower case character, alphabetically between a and n
*   - `[^arn]`
    - Returns a match for any character EXCEPT a, r, and n
*   - `[0123]`
    - Returns a match where any of the specified digits (0, 1, 2, or 3) are present
*   - `[0-9]`
    - Returns a match for any digit between 0 and 9
*   - `[0-5][0-9]`
    - Returns a match for any two-digit numbers from 00 and 59
*   - `[a-zA-Z]`
    - Returns a match for any character alphabetically between a and z, lower case OR upper case	
*   - `[+]`
    - In sets, +, *, ., |, (), $,{} has no special meaning, so [+] means: return a match for any + character in the string
:::

### The findall() Function

The findall() function returns a list containing all matches.

Example

Print a list of all matches:

In [18]:
import re

txt = "The rain in Korea"
x = re.findall("ai", txt)
print(x)

['ai', 'ai']


The list contains the matches in the order they are found.

If no matches are found, an empty list is returned:

Example

Return an empty list if no match was found:

In [19]:
import re

txt = "The rain in Korea"
x = re.findall("Portugal", txt)
print(x)

[]


### The `search()` Function

The `search()` function searches the string for a match, and returns a Match object if there is a match.

If there is more than one match, only the first occurrence of the match will be returned:

Example

Search for the first white-space character in the string:

In [57]:
import re

txt = "The rain in Korea"
x = re.search(r"\s", txt)

print("The first white-space character is located in position:", x.start())

The first white-space character is located in position: 3


If no matches are found, the value None is returned:

Example

Make a search that returns no match:

In [21]:
import re

txt = "The rain in Korea"
x = re.search("Portugal", txt)
print(x)

None


### The split() Function

The split() function returns a list where the string has been split at each match:

Example

Split at each white-space character:

In [56]:
import re

txt = "The rain in Korea"
x = re.split(r"\s", txt)
print(x)

['The', 'rain', 'in', 'Korea']


You can control the number of occurrences by specifying the maxsplit parameter:

Example

Split the string only at the first occurrence:

In [58]:
import re

txt = "The rain in Korea"
x = re.split(r"\s", txt, 1)
print(x)

['The', 'rain in Korea']


### The sub() Function

The sub() function replaces the matches with the text of your choice:

Example

Replace every white-space character with the number 9:

In [59]:
import re

txt = "The rain in Korea"
x = re.sub(r"\s", "9", txt)
print(x)

The9rain9in9Korea


You can control the number of replacements by specifying the count parameter:

Example

Replace the first 2 occurrences:

In [60]:
import re

txt = "The rain in Korea"
x = re.sub(r"\s", "9", txt, 2)
print(x)

The9rain9in Korea


### Match Object

A Match Object is an object containing information about the search and the result.

Note: If there is no match, the value None will be returned, instead of the Match Object.

Example

Do a search that will return a Match Object:

In [28]:
txt = "The rain in Korea"
x = re.search("ai", txt)
print(x) #this will print an object

<re.Match object; span=(5, 7), match='ai'>


The Match object has properties and methods used to retrieve information about the search, and the result:

- `.span()` returns a tuple containing the start-, and end positions of the match.
- `.string` returns the string passed into the function
- `.group()` returns the part of the string where there was a match

Example

Print the position (start- and end-position) of the first match occurrence.

The regular expression looks for any words that starts with an upper case "S":

In [30]:
txt = "The rain in Korea"
x = re.search(r"\bS\w+", txt)
print(x.span())

(12, 17)


Example

Print the string passed into the function:

In [31]:
txt = "The rain in Korea"
x = re.search(r"\bS\w+", txt)
print(x.string)

The rain in Spain


Example

Print the part of the string where there was a match.

The regular expression looks for any words that starts with an upper case "S":

In [14]:
txt = "The rain in Korea"
x = re.search(r"\bS\w+", txt)
print(x.group())

Spain


**연습**

Consider the following code:

In [15]:
txt = 'The rain in Korea'
x = re.findall('[a-c]', txt)
print(x)

['a', 'a']


**참고**

- [(W3 Schools) Python 정규표현식](https://www.w3schools.com/python/python_정규표현식.asp)
- [정규표현식One](https://정규표현식one.com)

## 예제

**예제 ?**

다음 세 개의 함수는 모두 입력된 문자열이 소문자를 포함하고 있는지 여부를 조사하도록 구현되었지만
하나만 제대로 작동한다.
어느 함수가 제대로 작동하는지 확인한 다음 그 이유를 설명하라.
또한 다른 두 함수가 적절하게 구현되지 않은 이유를 설명하라.

In [56]:
def any_lowercase1(text):
    for c in text:
        if c.islower():
            return True
        else:
            return False

def any_lowercase2(text):
    for c in text:
        if 'c'.islower():
            return 'True'
        else:
            return 'False'

def any_lowercase3(text):
    flag = False
    for c in text:
        flag = flag or c.islower()
    return flag

위 코드에 사용된 `islower()` 메서드는 지정된 문자열이 모두 대문자 알파벳을 포함하지 않는지 여부를 판단한다.
즉, 소문자와 기타 기호만 포함할 때 `True`를 반환한다.

In [57]:
'abc'.islower()

True

In [58]:
'Abc'.islower()

False

In [59]:
'abc !'.islower()

True

`isupper()` 메서드는 반대로 판단한다.

In [60]:
'aBC'.isupper()

False

In [61]:
'ABC'.isupper()

True

In [62]:
'ABC !'.isupper()

True

답:

아래 문자열에 대해 위 함수를 테스트한다.

In [63]:
word1 = "ABcDE"
word2 = "ABCDE"

**`any_lowercase1()` 함수**

In [64]:
def any_lowercase1(text):
    for c in text:
        if c.islower():
            return True
        else:
            return False

인자로 입력되는 문자열의 첫 문자가 소문자의 대소문자 여부를 확인하고 바로 반환값을 지정한다.
함수는 반환값이 지정되면 바로 실행을 종료하기에 결국 첫 문자만 확인하고 종료한다.
따라서 `word1` 변수가 가리키는 문자열에 소문자 `'c'`가 포함되어 있지만 첫 문자가 대문자이기에 `False`를 반환한다.

In [65]:
any_lowercase1(word1)

False

**`any_lowercase2()` 함수**

In [66]:
def any_lowercase2(text):
    for c in text:
        if 'c'.islower():
            return 'True'
        else:
            return 'False'

`for` 반복문의 본문이 함수의 인자를 전혀 활용하지 않는다.
이유는 `if 'c'.islower()` 에서 사용된 `'c'`는
`for c in text`의 `c`와 아무 상관이 없는 영어 알파벳 c 하나로
구성된 문자열이기 때문이다.
또한 `'c'`가 소문자이기에 결과는 항상 문자열 `'True'`로
심지어 부울값 `True`와 아무 상관이 없다.

In [67]:
any_lowercase2(word1)

'True'

In [68]:
any_lowercase2(word2)

'True'

**`any_lowercase3()` 함수**

In [69]:
def any_lowercase3(text):
    flag = False
    for c in text:
        flag = flag or c.islower()
    return flag

`flag` 변수는 소문자를 찾았는지 여부를 기억한다.
`for c in text` 반복문에서 `c`가 `text`에 포함된 문자 하나하나를 가리킬 때
소문자를 만나는 순간 `c.lower()`가 `True`가 되고 따라서 `or` 연산자에 의해
`flag`이 계속해서 `True`를 가리키게 된다.
즉, 소문자가 하나라도 포함되면 `True`를 반환한다.

In [70]:
any_lowercase3(word1)

True

In [71]:
any_lowercase3(word2)

False

결국 `any_lowercase3()` 만 소문자 포함여부를 제대로 결정한다.

**예제 ?**

변수와 함수의 이름을 지정할 때 **낙타**<font size = "2">camel</font> 표기법 또는
**팟홀**<font size="2">pothole</font> 표기법을 사용한다.

* 낙타 표기법: 소문자로 시작하고, 이어지는 단어의 시작은 대문자로 작성하는 표기법.
* 팟홀 표기법: 모두 소문자를 사용하고, 단어 사이에 밑줄기호(`_`)를 사용하는 표기법.

활용 예제는 다음과 같다.

- 낙타 표기법 예제: `userName`, `printMessage`, `countA` 등.
- 팟홀 표기법 예제: `user_name`, `print_message`, `count_a` 등.   

낙타 표기법을 팟홀 표기법으로 변환하는 `camel2pothole()` 함수를 구현하라.

힌트: 문자열 `isupper()`메서드 활용

답:

아래 함수는 주어진 문자열에 포함된 알파벳을 대상으로
소문자인 경우는 그대로,
대문자인 경우엔 소문자로 변환한 다음 밑줄과 함께 연결하는
방식으로 새로운 문자열을 생성한다.
해당 문자가 대문자인지 여부는 문자열 `isupper()` 메서드로 판단한다.

In [73]:
def camel2pothole(word):

    # 새 문자열 생성
    pothole_word = ""
    for char in word:
        if char.isupper():
            pothole_word += "_" + char.lower()
        else:
            pothole_word += char

    return pothole_word

In [74]:
print(camel2pothole("userName"))
print(camel2pothole("printMessage"))
print(camel2pothole("countA"))
print(camel2pothole("oneTwoThree"))

user_name
print_message
count_a
one_two_three


**예제 5**

'기러기', '토마토', '우영우', '인싸 의사의 싸인', 'bob', 'eye', 'pop'와 같이
앞으로 읽어도, 뒤로 읽어도 동일한 문자열을 **회문**<font size='2'>palindrome</font>이라 부른다.

(1) 문자열이 주어졌을 때 아래에 언급된 모든 기호가 제거된 문자열의 회문 여부를 확인하려 한다.

In [102]:
nonWord_chr = ' .,!?'

아래 문자열에서 앞서 언급된 기호를 모두 제거하면 회문이 됨을 보이는 코드를 작성하라.

힌트: 문자열 `replace()`

In [101]:
word = "여보게, 저기 저게 보여?"

답:

In [88]:
nonWord_chr = ' .,!?'
word = "여보게, 저기 저게 보여?"

for chr in nonWord_chr:
    word = word.replace(chr, "")

word

'여보게저기저게보여'

In [89]:
def no_nonWord(word):
    nonWord_chr = ' .,!?'

    for chr in nonWord_chr:
        word = word.replace(chr, "")

    return word

In [90]:
no_nonWord("여보게, 저기 저게 보여?")

'여보게저기저게보여'

In [91]:
no_nonWord("다시! 합창합시다.")

'다시합창합시다'

**예제 ?**

문자열의 `find()` 메서드가 작동하는 방식을 이해하기 위해 `find()` 함수를 직접 정의하려 한다.

**질문 1**

문자열 `find()` 메서드에 대응하는 `find()` 함수를 직접 정의하라.

* 매개 변수: 두 개의 문자열을 인자로 받는 `word1`과 `word2` 두 개 사용
* `word1`이 `word2`를 부분 문자열로 포함할 경우 부분 문자열이 시작하는 인덱스 반환.
    포함하지 않을 경우 -1을 반환.
    
단, 문자열 인덱싱, 슬라이싱만 사용하고 다른 문자열 메서드는 전혀 사용하지 않아야 한다.

답:

예를 들어 `'banana'` 문자열은 `'ana'` 문자열을 부분 문자열로 포함한다.
`'ana'` 부분 문자열이 `'banana'` 문자열의 어느 인덱스에서 시작하는지를
확인하려면 모든 인덱스에 대해 그곳에서부터 `'ana'` 문자열의 시작하는지 여부를
슬라이싱으로 대조하면 된다.

예를 들어 1번 인덱스부터 `'ana'` 문자열이 시작하는지 여부는 다음과 같이 확인한다.
아래 코드에서 3은 `'ana'` 문자열의 길이를 나타낸다.

```python
'banana'[1:1+3] == 'ana'
```

아래 `find()` 함수는 `word1` 문자열이 `word2` 문자열을 부분 문자열로 포함되는지 여부를
판단하기 위해 `word1`의 모든 인덱스에 대해 `word2` 부분 문자열과 일치 여부를
슬라이싱을 이용하여 확인하는 순간에 해당 인덱스를 반환한다.
또한 부분 문자열을 확인하지 못하면 -1을 반환한다.

In [75]:
def find(word1, word2):
    for idx in range(len(word1)):
        if word2 == word1[idx:idx+len(word2)] :
            return idx
    return -1

In [76]:
print(find('banana', 'ana'))
print(find('heyheyhey', 'ey'))
print(find('waikiki', 'ki'))
print(find('ulala', 'la'))
print(find('chukachuka', 'uKa'))

1
1
3
1
-1


**질문 2**

부분 문자열의 포함 여부를 지정된 인덱스부터 확인할 수 있도록 위 함수를 수정하라.
이를 위해 다음 `position` 매개변수를 셋째 매개변수로 추가한다.

* `position` 매개변수: 탐색을 시작하는 인덱스 지정. 기본 키워드 인자로 0 사용.

힌트: `range()` 함수의 구간 변경

답:

`find()` 함수의 셋째 매개변수로 `position=0`을 사용한다.
즉, 함수의 헤더를 다음과 같이 지정한다.

```python
def find(word1, word2, position=0)
```

`position=0`은 부분 문자열 탐색을 시작하는 인덱스를 가리키며 기본값으로 0을 사용한다.
즉, `position` 값을 지정하지 않거나 변경하지 않는다면 `word1`이 가리키는
문자열의 전 구간을 대상으로 부분 문자열 탐색 여부를 확인한다.

반면에 `position` 값을 0이 아닌 다른 값으로 지정하면
해당 인덱스부터 탐색을 시작한다.
이를 위해 `range()` 함수가 지정하는 아래처럼 구간을 `position`으로 시작하도록 한다.

```python
range(position, len(word1))
```

`find()` 함수의 다른 부분은 동일하며, 정리하면 다음과 같다.

In [77]:
def find(word1, word2, position=0):
    for idx in range(position, len(word1)):
        if word2 == word1[idx:idx+len(word2)] :
            return idx
    return -1

- `position=0`을 사용하는 경우

In [78]:
print(find('banana', 'ana'))
print(find('heyheyhey', 'ey'))
print(find('waikiki', 'ki'))
print(find('ulala', 'la'))
print(find('chukachuka', 'uKa'))

1
1
3
1
-1


- `position`을 다르게 지정하는 경우

In [79]:
print(find('banana', 'ana', 2))
print(find('heyheyhey', 'ey', 3))
print(find('waikiki', 'ki', 4))
print(find('ulala', 'la', 2))
print(find('chukachuka', 'uKa'))

3
4
5
3
-1


## 연습문제

참고: [(연습) 문자열 3부](https://colab.research.google.com/github/codingalzi/pybook/blob/master/practices/practice-strings_part3.ipynb)